File: tests/llvm/hardlift/benchmark/cpp/color_dodge_8.cc
#include <iostream>
#include <vector>
#include <chrono>

#include "utils.h"

using namespace std;
using namespace std::chrono;

vector<vector<float>> color_dodge_8(vector<vector<float>> base, vector<vector<float>> active) {
    vector<vector<float>> out;
    int m = base.size();
    int n = base[0].size();
    for (int row = 0; row < m; row++) {
        vector<float> row_vec;
        for (int col = 0; col < n; col++) {
            float pixel;
            if (active[row][col] == 255)
                pixel = 255;
            else
                pixel = base[row][col] / (255 - active[row][col]);
            row_vec.push_back(pixel);
        }
        out.push_back(row_vec);
    }
    return out;
}


int main() {
    mat_timer(&color_dodge_8);
    return 0;
}

File: tenspiler/blend/python/for_synthesis/color_dodge_8.py
from typing import List

from tenspiler.llm.python_dsl import (
    matrix_elemwise_div,
    matrix_scalar_sub,
    matrix_where,
)


def color_dodge_8_baseline(
    base: List[List[int]], active: List[List[int]]
) -> List[List[int]]:
    out = []
    m = len(base)
    n = len(base[0])

    for row in range(m):
        row_vec = []
        for col in range(n):
            if active[row][col] == 32:
                pixel = 32
            else:
                pixel = base[row][col] // (32 - active[row][col])
            row_vec.append(pixel)
        out.append(row_vec)

    return out


def color_dodge_8(base: List[List[int]], active: List[List[int]]) -> List[List[int]]:
    return matrix_where(
        active,
        matrix_elemwise_div(base, matrix_scalar_sub(32, active)),
        lambda a, b: 32 if a == 32 else b,
    )


active = [[123, 23, 46], [98, 98, 58]]
base = [[234, 87, 65], [10, 17, 70]]
print(color_dodge_8(base, active))
print(color_dodge_8_baseline(base, active))
